File: 07-Sorting/SelectionSort.java
package com.Anwesh;

import java.util.Arrays;

public class SelectionSort {
    public static void main(String[] args) {
        int[] arr = {4, 5, 1 ,2, 3};
        Sort(arr);
        System.out.println(Arrays.toString(arr));
    }
    static void Sort(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            int last = arr.length - i - 1;
            int maxIndex = getMaxIndex(arr, 0, last);
            swap(arr, maxIndex, last);
        }
    }
    static int getMaxIndex(int[] arr, int start, int end) {
        int max = start;
        for (int i = start; i <= end; i++) {
            if (arr[max] < arr[i])
                max = i;
        }
        return max;
    }
    static void swap(int[] arr, int first, int second) {
        int temp = arr[first];
        arr[first] = arr[second];
        arr[second] = temp;
    }
}

File: 07-Sorting/C/selection_sort.c
#include <stdio.h>
#include <stdlib.h>

int max(int *vetor, int start, int end) {
  int bigger = end;

  for (int i = start; i <= end; i++) {
    if (vetor[i] > vetor[end] && vetor[i] > vetor[bigger]) {
      bigger = i;
    }
  }

  return bigger;
}

void swap(int *vetor, int bigger, int lastPosition) {
  int temp = vetor[lastPosition];
  vetor[lastPosition] = vetor[bigger];
  vetor[bigger] = temp;
}

void selectionSort(int *vetor, int size) {
  int bigger;
  for (int i = size - 1; i > 0; i--) {
    bigger = max(vetor, 0, i);
    swap(vetor, bigger, i);
  }
}

void printVetor(int *vetor, int size) {
  for (int i = 0; i < size; i++) {
    printf("%d, ", vetor[i]);
  }
}

int main(int argc, char **argv) {
  int vet[5] = {-2, 6, 4, 10, 3};

  selectionSort(vet, 5);
}
